frontend_tests/zjsunit/namespace.js

```diff
@@ -19,28 +21,57 @@ exports.patch_builtin = function (name, val) {
     return val;
 };
 
-exports.zrequire = function (name, fn) {
+function resolve_name(name, fn) {
     if (fn === undefined) {
         fn = '../../static/js/' + name;
     } else if (/generated\/|js\/|third\//.test(fn)) {
         // FIXME: Stealing part of the NPM namespace is confusing.
         fn = '../../static/' + fn;
     }
-    delete require.cache[require.resolve(fn)];
-    var obj = require(fn);
-    requires.push(fn);
+    return fn;
+}
+
+// Require a module and get a handle to it for monkey patching.
+// Can not patch if module.exports is not an object (only objects are passed by
+// reference) or the module is imported with ES6's import statement (it has a
+// separate cache).
+// Put this before other zrequire-s if you run into caching issues.
+exports.zrequire_pure = function (name, fn) {
+    fn = resolve_name(name, fn);
+    const path = require.resolve(fn);
+    require(fn);
+    return require.cache[path].exports;
+};
+
+exports.zrequire = function (name, fn) {
+    var obj = exports.zrequire_pure(name, fn);
     set_global(name, obj);
```

Comment:
Since most of our “modules” already insert themselves into the global object, I wonder if most of our `zrequire`s can already be switched to `zrequire_pure`, in which case we could just call `zrequire_pure` `zrequire` and inline the `set_global` around the other calls as necessary?

Reply:
That doesn't work, the IIFE modules are inserted into `window` and not `global`. There were some old discussion in NodeJS on making them the same object but never got implemented.

Reply:
We…have that power, right?
```js
window = global;
```

Reply:
:smile: I will do that in another PR, don't want to bloat this up too much.

Reply:
@andersk modules aren't all loaded in node tests like `bundles/app.js` (and they shouldn't be), so this still wouldn't work.

Reply:
It does work: #12896.

Reply:
Ah sorry I misread you. Yes let's do this.

Reply:
@andersk can I add your commits to this PR?